zstd_stream.go

```diff
@@ -447,12 +448,12 @@ func (r *reader) Read(p []byte) (int, error) {
 	// at least one zstd block, so that we don't block if the
 	// other end has flushed a block.
 	for {
-		// - If the last decompression didn't entirely fill the decompression buffer,
-		//   zstd flushed all it could, and needs new data. In that case, do 1 Read.
-		// - If the last decompression did entirely fill the decompression buffer,
-		//   it might have needed more room to decompress the input. In that case,
-		//   don't do any unnecessary Read that might block.
-		needsData := r.decompSize < len(r.decompressionBuffer)
+		// Do a read only when zstd doesn't have any pending output. When zstd could
+		// not finish writing output, it will keep input buffer non-empty.
+		// https://github.com/facebook/zstd/commit/b3060f7a9ea3555c6045606be58dddc86bbb099b
+		// > when all compressed frame is consumed, it means decompression is completed,
+		// > with regenerated data fully flushed.
+		needsData := r.compressionLeft == 0
```

Comment:
I was worried about this change potentially inducing an infinite loop (zstd waiting for addl data but consuming 0 from the input buffer), however I did run fuzzing on this endpoint and did not find anything so g2g